packages/core/src/execution-engine/workflow-execute.ts

```diff
@@ -1572,7 +1573,14 @@ export class WorkflowExecute {
 								if (runNodeData.closeFunction) {
 									// Explanation why we do this can be found in n8n-workflow/Workflow.ts -> runNode
 
-									closeFunction = runNodeData.closeFunction();
+									try {
+										closeFunction = runNodeData.closeFunction();
+									} catch (error) {
+										throw new TriggerCloseError(executionNode, {
+											cause: error as Error,
+											level: 'warning',
+										});
+									}
```

Comment:
Can you help me understand how this change ensures that a deactivation error does not block deactivation? As here we are executing rather than deactivating?

Reply:
This is correct. This is also the [same thing that we do in `ActiveWorklows`](https://github.com/n8n-io/n8n/blob/master/packages/core/src/execution-engine/active-workflows.ts#L233).

Logic duplication like this makes me wonder if we need a `WorkflowLifecycle` like we have execution lifecycle, so that we can abstract all activation and deactivation code into a centralized place.

Reply:
found another issue in this that `closeFunction` is async, which means that this won't throw until await is called on the returned promise.
I'll have a look.